### PR TITLE
feat: laser encoder

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
https://github.com/jina-ai/internal-tasks/issues/92

### CODE REVIEW

File: laser_encoder.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        from laserembeddings import Laser
+        self._path_to_bpe_codes = path_to_bpe_codes or Laser.DEFAULT_BPE_CODES_FILE
+        self._path_to_bpe_vocab = path_to_bpe_vocab or Laser.DEFAULT_BPE_VOCAB_FILE
+        self._path_to_encoder = path_to_encoder or Laser.DEFAULT_ENCODER_FILE
+        self.on_gpu = on_gpu
+        self.default_batch_size = default_batch_size
+        self.default_traversal_paths = default_traversal_paths
+        self.language = language.lower()
+
+        from laserembeddings import Laser
```
Comment (JoanFM) [inline laser_encoder.py]: why twice? Lets import it on the top, no need to hide here anymore

File: tests/integration/test_exec.py [hunk 2]
```diff
@@ -1,12 +1,54 @@
...
+        result = f.post(on='/test', inputs=docs, return_results=True)
+        for path, count in [['r', 0], ['c', 3], ['cc', 0]]:
+            assert len(DocumentArray(result[0].data.docs).traverse_flat([path]).get_attributes('embedding')) == count
+
+        result = f.post(on='/test', inputs=docs, parameters={'traversal_paths': ['cc']}, return_results=True)
+        for path, count in [['r', 0], ['c', 0], ['cc', 2]]:
+            assert len(DocumentArray(result[0].data.docs).traverse_flat([path]).get_attributes('embedding')) == count
+
+
+def test_no_documents():
```
Comment (JoanFM) [inline tests/integration/test_exec.py]: This test should be in unit

File: tests/integration/test_exec.py [hunk 3]
```diff
@@ -1,12 +1,54 @@
...
+
+def test_use_in_flow():
+    with Flow.load_config('flow.yml') as flow:
+        data = flow.post(on='/encode', inputs=data_generator(5))
+        docs = data[0].docs
+        for doc in docs:
+            assert doc.embedding.shape == (1024,)
+
+
+def test_traversal_path():
```
Comment (JoanFM) [inline tests/integration/test_exec.py]: this one as well

File: tests/integration/test_exec.py [hunk 4]
```diff
@@ -1,12 +1,54 @@
...
+    for i in range(num_docs):
+        doc = Document(
+            text='it is a good day! the dog sits on the floor.')
+        yield doc
+
+
+def test_use_in_flow():
+    with Flow.load_config('flow.yml') as flow:
+        data = flow.post(on='/encode', inputs=data_generator(5))
+        docs = data[0].docs
```
Comment (JoanFM) [inline tests/integration/test_exec.py]: assert length of docs, otherwise an empty list would pass the test

**Review by JoanFM — DISMISSED** [review]

File: laser_encoder.py (hunk 1 above)
Comment (Yongxuanzhang) [reply laser_encoder.py]: done

File: tests/integration/test_exec.py (hunk 2 above)
Comment (Yongxuanzhang) [reply tests/integration/test_exec.py]: done

File: tests/integration/test_exec.py (hunk 3 above)
Comment (Yongxuanzhang) [reply tests/integration/test_exec.py]: done

File: tests/integration/test_exec.py (hunk 4 above)
Comment (Yongxuanzhang) [reply tests/integration/test_exec.py]: done

File: Dockerfile [hunk 5]
```diff
@@ -5,7 +5,7 @@ RUN apt-get -y update && apt-get install -y git
 
 # install requirements before copying the workspace
 COPY requirements.txt /requirements.txt
-RUN pip install -r requirements.txt
+RUN pip install -r requirements.txt && python -m laserembeddings download-models
```
Comment (cristianmtr) [inline Dockerfile:10]: not running tests in Dockerfile?

File: laser_encoder.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    """
+
+    def __init__(
+            self,
+            path_to_bpe_codes: Optional[str] = None,
+            path_to_bpe_vocab: Optional[str] = None,
+            path_to_encoder: Optional[str] = None,
+            on_gpu: bool = False,
+            default_batch_size: int = 32,
+            default_traversal_paths: List[str] = 'r',
```
Comment (cristianmtr) [inline laser_encoder.py]:
```suggestion
            default_traversal_paths: List[str] = ['r'],
```

File: tests/unit/test_exec.py [hunk 7]
```diff
@@ -1,9 +1,50 @@
...
+                      ]
+    docs[0].chunks[0].chunks = [
+        Document(id='chunk111', text=text),
+        Document(id='chunk112', text=text),
+    ]
+
+    encoder = LaserEncoder()
+    encoder.encode(docs, parameters={'batch_size': 10, 'traversal_paths': ['c']})
+
+    for path, count in [['r', 0], ['c', 3], ['cc', 0]]:
```
Comment (cristianmtr) [inline tests/unit/test_exec.py]: traversal_paths should be `['r']`, `['c']` -- `List[str]`

File: tests/unit/test_exec.py [hunk 8]
```diff
@@ -1,9 +1,50 @@
...
+        assert len(docs.traverse_flat([path]).get_attributes('embedding')) == count
+
+    encoder.encode(docs, parameters={'batch_size': 10, 'traversal_paths': ['cc']})
+    for path, count in [['r', 0], ['c', 3], ['cc', 2]]:
+        assert len(docs.traverse_flat([path]).get_attributes('embedding')) == count
```
Comment (cristianmtr) [inline tests/unit/test_exec.py]: maybe add an assert on the shape of embeddings

**Review by cristianmtr — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 5 above)
Comment (Yongxuanzhang) [reply Dockerfile:10]: I don't see any other new executors running test in the docker file

File: laser_encoder.py (hunk 6 above)
Comment (Yongxuanzhang) [reply laser_encoder.py]: fixed

File: tests/unit/test_exec.py (hunk 8 above)
Comment (Yongxuanzhang) [reply tests/unit/test_exec.py]: added

File: tests/unit/test_exec.py (hunk 7 above)
Comment (Yongxuanzhang) [reply tests/unit/test_exec.py]: changed

File: Dockerfile (hunk 5 above)
Comment (nan-wang) [reply Dockerfile:10]: > not running tests in Dockerfile?
Nope. Hubble is not running tests when building images.

File: README.md [hunk 9]
```diff
@@ -21,7 +17,9 @@
 
 ## 🌱 Prerequisites
 
-Some conditions to fulfill before running the executor
+```bash
+python -m laserembeddings download-models
```
Comment (nan-wang) [inline README.md:21]:
```suggestion
pip install -U -r requirements.txt
python -m laserembeddings download-models
```

File: manifest.yml [hunk 10]
```diff
@@ -1,8 +1,8 @@
...
-description: This is my dummy executor
+name: LaserEncoder
+alias: LaserEncoder
+description: encodes data from an 1d array of string in size `B` into an ndarray in size `B x D`.
 author: Jina AI Dev-Team (dev-team@jina.ai)
-url: https://github.com/jina-ai/EXECUTOR_REPO_NAME
+url: https://github.com/jina-ai/executor-text-laser-encoder
 avatar: https://avatars.githubusercontent.com/u/60539444?s=200&v=4
-keywords: [encoder, Numpy]
+keywords: [nlp, encoder, LASER, multilingual]
```
Comment (Kelton8Z) [inline manifest.yml]: according to the guideline, model name should be lowercased, and nlp should be text

Comment (Yongxuanzhang) [reply manifest.yml]: got it, thanks!

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Review by cristianmtr — APPROVED** [review]

**Comment by cristianmtr** [comment]
CI is failing though

**Comment by Yongxuanzhang** [comment]
> CI is failing though

Yes I will fix it before merging

**Comment by nan-wang** [comment]
> Yes I will fix it before merging

I've fixed it

**Pull request merged**